File: sdk/desktop/windows/tests/Kseal.Desktop.Tests/TrustCoreTests.cs
using Xunit;

namespace Kseal.Desktop.Tests;

/// <summary>
/// Exercises the real Rust core over the C ABI (no faked core). Requires the
/// shared library on the load path — KSEAL_FFI_LIBRARY points at it (see
/// scripts/build-rust-host.sh). These tests run the same code path production
/// uses.
/// </summary>
public class TrustCoreTests
{
    private static NativeTrustCore NewCore() =>
        NativeTrustCore.Create(new byte[32], new byte[32]);

    [Fact]
    public void VersionIsNonEmpty()
    {
        using var core = NewCore();
        Assert.False(string.IsNullOrEmpty(core.Version));
    }

    [Fact]
    public void NonceHasRequestedLength()
    {
        using var core = NewCore();
        Assert.Equal(16, core.GenerateNonce(16).Length);
        Assert.Equal(32, core.GenerateNonce(32).Length);
    }

    [Fact]
    public void NoncesAreUnique()
    {
        using var core = NewCore();
        byte[] a = core.GenerateNonce(16);
        byte[] b = core.GenerateNonce(16);
        Assert.NotEqual(a, b);
    }

    [Fact]
    public void CompressDecompressRoundTrips()
    {
        using var core = NewCore();
        byte[] data = System.Text.Encoding.UTF8.GetBytes(new string('k', 1024));
        byte[] compressed = core.Compress(data, 0);
        Assert.True(compressed.Length < data.Length);
        Assert.Equal(data, core.Decompress(compressed));
    }

    [Fact]
    public void CleanBitsScoreZero()
    {
        using var core = NewCore();
        var score = core.EvaluateRisk(0);
        Assert.Equal(0u, score.Score);
    }

    [Fact]
    public void RiskBitsProduceNonZeroScore()
    {
        using var core = NewCore();
        ulong bits = RiskSignals.Pack([RiskSignal.Tamper, RiskSignal.AppIntegrity]);
        Assert.True(core.EvaluateRisk(bits).Score > 0);
    }

    [Fact]
    public void TrustLevelUnspecifiedWithoutPolicy()
    {
        using var core = NewCore();
        ulong bits = RiskSignal.Tamper.Mask();
        Assert.Equal(TrustLevel.Unspecified, core.ComputeRiskLevel(bits));
    }

    [Fact]
    public void EvaluateRiskAndLevelAgreesWithSeparateCalls()
    {
        using var core = NewCore();
        ulong bits = RiskSignals.Pack([RiskSignal.Hooking, RiskSignal.Debugger]);
        var (score, level) = core.EvaluateRiskAndLevel(bits);
        Assert.Equal(core.EvaluateRisk(bits).Score, score.Score);
        Assert.Equal(core.ComputeRiskLevel(bits), level);
    }

    [Fact]
    public void CreateEventAndBatchProduceWire()
    {
        using var core = NewCore();
        byte[] evt = core.CreateEvent(
            EventType.RuntimeTamper, RiskSignal.Tamper.Mask(), Confidence.High,
            buildHash: "build-abc", policyHash: "policy-xyz", installKeyHash: "install-1",
            coarseTimeBucket: 0, country: null);
        Assert.NotEmpty(evt);

        byte[] wire = core.BatchAndCompress([evt]);
        Assert.NotEmpty(wire);
    }

    [Fact]
    public void RequestProofIsDeterministicAndSequenceSensitive()
    {
        using var core = NewCore();
        byte[] requestHash = System.Text.Encoding.UTF8.GetBytes("request-hash");
        byte[] nonce = core.GenerateNonce(16);

        byte[] proof1 = core.GenerateRequestProof("token-1", requestHash, nonce, 1);
        byte[] proof1Again = core.GenerateRequestProof("token-1", requestHash, nonce, 1);
        byte[] proof2 = core.GenerateRequestProof("token-1", requestHash, nonce, 2);

        Assert.Equal(proof1, proof1Again);   // deterministic for identical inputs
        Assert.NotEqual(proof1, proof2);      // sequence number changes the proof
    }

    [Fact]
    public void VerifyConfigSignatureRejectsGarbage()
    {
        Assert.False(NativeTrustCore.VerifyConfigSignature(new byte[8], new byte[64], new byte[32]));
    }

    [Fact]
    public void DisposeIsIdempotent()
    {
        var core = NewCore();
        core.Dispose();
        // A second dispose must be a no-op, not throw on the freed lock.
        var ex = Record.Exception(() => core.Dispose());
        Assert.Null(ex);
    }
}

File: sdk/desktop/macos/Tests/KsealDesktopTests/TrustCoreTests.swift
import XCTest
@testable import KsealDesktop

/// Exercises the REAL Rust trust core through the C ABI on the host.
///
/// The shared library `libkseal_ffi.{so,dylib}` is built by
/// `scripts/build-rust-host.sh` and linked by `Package.swift`. This is not a
/// mock — it is the same `kseal_*` C ABI the desktop SDK ships, validating the
/// FFI integration end to end.
final class TrustCoreTests: XCTestCase {

    private func makeCore() throws -> NativeTrustCore {
        try NativeTrustCore.create(
            configPublicKey: Data(repeating: 7, count: 32),
            proofKey: Data("instance-key".utf8),
            platform: .desktopMac
        )
    }

    func testVersionIsNonEmpty() throws {
        XCTAssertFalse(try makeCore().version.isEmpty)
    }

    func testNonceHasRequestedLengthAndIsRandom() throws {
        let core = try makeCore()
        let n1 = try core.generateNonce(16)
        let n2 = try core.generateNonce(16)
        XCTAssertEqual(n1.count, 16)
        XCTAssertNotEqual(n1, n2)
    }

    func testCompressDecompressRoundTrips() throws {
        let core = try makeCore()
        let payload = Data(repeating: 0x41, count: 512)
        let compressed = try core.compress(payload, level: 0)
        XCTAssertFalse(compressed.isEmpty)
        XCTAssertEqual(try core.decompress(compressed), payload)
    }

    func testEvaluateRiskUsesDefaultWeights() throws {
        let core = try makeCore()
        let bits = RiskSignal.pack([.tamper, .hooking])
        let score = try core.evaluateRisk(bits)
        XCTAssertEqual(score.score, 20) // default per-signal weight of 10
        XCTAssertEqual(score.confidence, .medium)
    }

    func testCleanBitsScoreZero() throws {
        let core = try makeCore()
        let score = try core.evaluateRisk(0)
        XCTAssertEqual(score.score, 0)
        XCTAssertEqual(score.confidence, .high)
    }

    func testTrustLevelUnspecifiedWithoutPolicy() throws {
        let core = try makeCore()
        XCTAssertEqual(core.computeRiskLevel(RiskSignal.tamper.mask), .unspecified)
    }

    func testCreateEventAndBatchProduceWire() throws {
        let core = try makeCore()
        let event = try core.createEvent(
            eventType: .appIntegrityFail,
            riskBits: RiskSignal.appIntegrity.mask,
            confidence: .low,
            buildHash: "build",
            policyHash: "policy",
            installKeyHash: "install",
            coarseTimeBucket: 1_700_000_000,
            country: nil
        )
        XCTAssertFalse(event.isEmpty)
        XCTAssertFalse(try core.batchAndCompress([event]).isEmpty)
    }

    func testRequestProofIsDeterministicAndSequenceSensitive() throws {
        let core = try makeCore()
        let hash = Data((0..<32).map { UInt8($0) })
        let nonce = Data(repeating: 9, count: 16)
        let p1 = try core.generateRequestProof(tokenId: "tok-1", requestHash: hash, nonce: nonce, sequence: 7)
        let p2 = try core.generateRequestProof(tokenId: "tok-1", requestHash: hash, nonce: nonce, sequence: 7)
        let p3 = try core.generateRequestProof(tokenId: "tok-1", requestHash: hash, nonce: nonce, sequence: 8)
        XCTAssertFalse(p1.isEmpty)
        XCTAssertEqual(p1, p2)
        XCTAssertNotEqual(p1, p3)
    }
}
